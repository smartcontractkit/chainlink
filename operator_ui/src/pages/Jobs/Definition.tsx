import React from 'react'
import {
  createStyles,
  CardContent,
  Divider,
  Card,
  Grid,
  Theme,
  Typography,
  withStyles,
  WithStyles,
} from '@material-ui/core'
import Content from 'components/Content'
import PrettyJson from 'components/PrettyJson'
import { JobData } from './sharedTypes'

const definitionStyles = (theme: Theme) =>
  createStyles({
    definitionTitle: {
      marginTop: theme.spacing.unit * 2,
      marginBottom: theme.spacing.unit * 2,
    },
    divider: {
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit * 3,
    },
  })

const Definition: React.FC<
  {
    error: unknown
    ErrorComponent: React.FC
    LoadingPlaceholder: React.FC
    job?: JobData['job']
  } & WithStyles<typeof definitionStyles>
> = ({ classes, error, ErrorComponent, LoadingPlaceholder, job }) => {
  React.useEffect(() => {
    document.title = job?.name
      ? `${job.name} | Job definition`
      : 'Job definition'
  }, [job])

  return (
    <Content>
      <Card>
        <ErrorComponent />
        <LoadingPlaceholder />
        {!error && job && (
          <CardContent>
            <Grid container spacing={0}>
              <Grid item xs={12}>
                <Typography variant="h5" className={classes.definitionTitle}>
                  Definition
                </Typography>
              </Grid>
              <Grid item xs={12}>
                <Divider light className={classes.divider} />
              </Grid>
              <Grid item xs={12}>
                {job.type === 'Direct request' && (
                  <PrettyJson object={JSON.parse(job.definition)} />
                )}
                {job.type === 'Off-chain reporting' && (
                  <Typography
                    style={{ margin: 0 }}
                    variant="body1"
                    component="pre"
                  >
                    {job.definition}
                  </Typography>
                )}
              </Grid>
            </Grid>
          </CardContent>
        )}
      </Card>
    </Content>
  )
}

export const JobsDefinition = withStyles(definitionStyles)(Definition)
export default JobsDefinition
