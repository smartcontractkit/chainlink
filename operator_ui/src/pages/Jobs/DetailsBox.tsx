import React from 'react'

import Grid from '@material-ui/core/Grid'
import Paper from '@material-ui/core/Paper'
import {
  createStyles,
  Theme,
  withStyles,
  WithStyles,
} from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'

import { JobData } from './sharedTypes'
import { TimeAgo } from 'src/components/TimeAgo'

const styles = (theme: Theme) =>
  createStyles({
    paper: {
      margin: `${theme.spacing.unit * 2.5}px 0`,
      padding: theme.spacing.unit * 3,
    },
  })

interface Props extends WithStyles<typeof styles> {
  job: JobData['job']
}

export const DetailsBox = withStyles(styles)(({ classes, job }: Props) => {
  if (!job) {
    return null
  }

  return (
    <Paper className={classes.paper}>
      <Grid container>
        <Grid item xs={12} sm={6} md={1}>
          <Typography variant="subtitle2" gutterBottom>
            ID
          </Typography>
          <Typography variant="body1">{job.id}</Typography>
        </Grid>
        <Grid item xs={12} sm={6} md={2}>
          <Typography variant="subtitle2" gutterBottom>
            Type
          </Typography>
          <Typography variant="body1">{job.specType}</Typography>
        </Grid>
        <Grid item xs={12} sm={6} md={5}>
          <Typography variant="subtitle2" gutterBottom>
            External Job ID
          </Typography>
          <Typography variant="body1">{job.externalJobID}</Typography>
        </Grid>
        <Grid item xs={12} sm={6} md={4}>
          <Typography variant="subtitle2" gutterBottom>
            Created At
          </Typography>
          <Typography variant="body1">
            <TimeAgo tooltip>{job.createdAt}</TimeAgo>
          </Typography>
        </Grid>
      </Grid>
    </Paper>
  )
})
