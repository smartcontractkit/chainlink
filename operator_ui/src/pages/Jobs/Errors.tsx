import React from 'react'
import {
  Card,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Typography,
} from '@material-ui/core'
import { v2 } from 'api'
import Button from 'components/Button'
import Content from 'components/Content'
import { localizedTimestamp, TimeAgo } from '@chainlink/styleguide'
import { JobData } from './sharedTypes'

export const JobsErrors: React.FC<{
  error: unknown
  ErrorComponent: React.FC
  LoadingPlaceholder: React.FC
  job?: JobData['job']
  setState: React.Dispatch<React.SetStateAction<JobData>>
  getJobSpec: () => Promise<void>
}> = ({
  error,
  ErrorComponent,
  getJobSpec,
  LoadingPlaceholder,
  job,
  setState,
}) => {
  React.useEffect(() => {
    document.title = job?.name ? `${job?.name} | Job errors` : 'Job errors'
  }, [job])

  const handleDismiss = async (jobSpecErrorId: string) => {
    // Optimistic delete
    const jobCopy: NonNullable<JobData['job']> = JSON.parse(JSON.stringify(job))
    jobCopy.errors = jobCopy.errors.filter((e) => e.id !== jobSpecErrorId)
    setState((state) => ({ ...state, job: jobCopy }))

    await v2.jobSpecErrors.destroyJobSpecError(jobSpecErrorId)
    await getJobSpec()
  }

  const tableHeaders = ['Occurrences', 'Created', 'Last Seen', 'Message']

  if (job?.type === 'Direct request') {
    tableHeaders.push('Actions')
  }

  return (
    <Content>
      <ErrorComponent />
      <LoadingPlaceholder />
      {!error && job && (
        <Card>
          <Table>
            <TableHead>
              <TableRow>
                {tableHeaders.map((header) => (
                  <TableCell key={header}>
                    <Typography variant="body1" color="textSecondary">
                      {header}
                    </Typography>
                  </TableCell>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {job.errors.length === 0 ? (
                <TableRow>
                  <TableCell component="th" scope="row" colSpan={5}>
                    No errors
                  </TableCell>
                </TableRow>
              ) : (
                job.errors.map((jobSpecError) => (
                  <TableRow key={jobSpecError.id}>
                    <TableCell>
                      <Typography variant="body1">
                        {jobSpecError.occurrences}
                      </Typography>
                    </TableCell>
                    <TableCell>
                      <Typography variant="body1">
                        <TimeAgo tooltip>
                          {localizedTimestamp(
                            jobSpecError.createdAt.toString(),
                          )}
                        </TimeAgo>
                      </Typography>
                    </TableCell>
                    <TableCell>
                      <Typography variant="body1">
                        <TimeAgo tooltip>
                          {localizedTimestamp(
                            jobSpecError.updatedAt.toString(),
                          )}
                        </TimeAgo>
                      </Typography>
                    </TableCell>
                    <TableCell>
                      <Typography variant="body1">
                        {jobSpecError.description}
                      </Typography>
                    </TableCell>
                    {job?.type === 'Direct request' && (
                      <TableCell>
                        <Button
                          variant="danger"
                          size="small"
                          onClick={() => {
                            handleDismiss(jobSpecError.id)
                          }}
                        >
                          Dismiss
                        </Button>
                      </TableCell>
                    )}
                  </TableRow>
                ))
              )}
            </TableBody>
          </Table>
        </Card>
      )}
    </Content>
  )
}

export default JobsErrors
