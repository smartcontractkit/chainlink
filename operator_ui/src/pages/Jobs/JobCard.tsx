import React from 'react'

import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'

import { JobData } from './sharedTypes'
import { DetailsCard } from 'src/components/Cards/DetailsCard'
import { TimeAgo } from 'src/components/TimeAgo'

interface Props {
  job: JobData['job']
}

export const JobCard = ({ job }: Props) => {
  if (!job) {
    return null
  }

  return (
    <DetailsCard>
      <Grid container>
        <Grid item xs={12} sm={6} md={1}>
          <Typography variant="subtitle2" gutterBottom>
            ID
          </Typography>
          <Typography variant="body1">{job.id}</Typography>
        </Grid>
        <Grid item xs={12} sm={6} md={2}>
          <Typography variant="subtitle2" gutterBottom>
            Type
          </Typography>
          <Typography variant="body1">{job.specType}</Typography>
        </Grid>
        <Grid item xs={12} sm={6} md={5}>
          <Typography variant="subtitle2" gutterBottom>
            External Job ID
          </Typography>
          <Typography variant="body1">{job.externalJobID}</Typography>
        </Grid>
        <Grid item xs={12} sm={6} md={4}>
          <Typography variant="subtitle2" gutterBottom>
            Created At
          </Typography>
          <Typography variant="body1">
            <TimeAgo tooltip>{job.createdAt}</TimeAgo>
          </Typography>
        </Grid>
      </Grid>
    </DetailsCard>
  )
}
