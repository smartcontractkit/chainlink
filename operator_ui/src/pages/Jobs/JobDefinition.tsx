import React from 'react'
import {
  createStyles,
  CardContent,
  Card,
  Theme,
  Typography,
  withStyles,
  WithStyles,
} from '@material-ui/core'

import { CardTitle } from 'components/CardTitle'
import Content from 'components/Content'
import PrettyJson from 'components/PrettyJson'
import { JobData } from './sharedTypes'

const definitionStyles = (theme: Theme) =>
  createStyles({
    definitionTitle: {
      marginTop: theme.spacing.unit * 2,
      marginBottom: theme.spacing.unit * 2,
    },
    divider: {
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit * 3,
    },
  })

type Props = {
  error: unknown
  ErrorComponent: React.FC
  LoadingPlaceholder: React.FC
  job?: JobData['job']
} & WithStyles<typeof definitionStyles>

export const JobDefinition = withStyles(definitionStyles)(
  ({ error, ErrorComponent, LoadingPlaceholder, job }: Props) => {
    React.useEffect(() => {
      document.title = job?.name
        ? `${job.name} | Job definition`
        : 'Job definition'
    }, [job])

    return (
      <Content>
        <ErrorComponent />
        <LoadingPlaceholder />
        {!error && job && (
          <Card>
            <CardTitle divider>Definition</CardTitle>
            <CardContent>
              {job.type === 'Direct request' && (
                <PrettyJson object={JSON.parse(job.definition)} />
              )}
              {job.type === 'v2' && (
                <Typography
                  style={{ margin: 0 }}
                  variant="body1"
                  component="pre"
                >
                  {job.definition}
                </Typography>
              )}
            </CardContent>
          </Card>
        )}
      </Content>
    )
  },
)
