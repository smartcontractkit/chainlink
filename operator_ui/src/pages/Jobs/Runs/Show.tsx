import React from 'react'
import { v2 } from 'api'
import { Route, RouteComponentProps, Switch } from 'react-router-dom'
import { CardTitle } from '@chainlink/styleguide'
import { Card, Grid, Typography } from '@material-ui/core'
import Content from 'components/Content'
import StatusCard from './StatusCard'
import { useErrorHandler } from 'hooks/useErrorHandler'
import { useLoadingPlaceholder } from 'hooks/useLoadingPlaceholder'
import RegionalNav from './RegionalNav'
import { DirectRequestJobRun } from '../sharedTypes'
import { Overview } from './Overview/Overview'
import { Json } from './Json'

type Props = RouteComponentProps<{
  jobSpecId: string
  jobRunId: string
}>

export const Show = ({ match }: Props) => {
  const [jobRun, setState] = React.useState<DirectRequestJobRun>()

  const { jobSpecId, jobRunId } = match.params

  const { error, ErrorComponent, setError } = useErrorHandler()
  const { LoadingPlaceholder } = useLoadingPlaceholder(!error && !jobRun)

  React.useEffect(() => {
    document.title = 'Job run details'
  }, [])

  const getJobRun = React.useCallback(async () => {
    return v2.runs
      .getJobSpecRun(jobRunId)
      .then((jobSpecRunsResponse) => {
        const jobRun = jobSpecRunsResponse.data
        setState({
          ...jobSpecRunsResponse.data.attributes,
          id: jobRun.id,
          jobId: jobSpecId,
        })
      })
      .catch(setError)
  }, [jobRunId, jobSpecId, setError])

  React.useEffect(() => {
    getJobRun()
  }, [getJobRun])

  const erroredTasks = jobRun
    ? jobRun.taskRuns.filter((tr) => tr.status === 'errored')
    : []

  return (
    <>
      <RegionalNav {...match.params} jobRun={jobRun} />
      <Content>
        <ErrorComponent />
        <LoadingPlaceholder />
        {jobRun && (
          <Grid container spacing={40}>
            <Grid item xs={8}>
              <Grid container spacing={40}>
                {erroredTasks.length > 0 && (
                  <Grid item xs={12}>
                    <Card>
                      <CardTitle divider>Errors</CardTitle>
                      <ul>
                        {erroredTasks.map((tr) => (
                          <li key={tr.id}>
                            <Typography variant="body1">
                              {tr.result.error}
                            </Typography>
                          </li>
                        ))}
                      </ul>
                    </Card>
                  </Grid>
                )}
                <Grid item xs={12}>
                  <Switch>
                    <Route
                      path={`${match.path}/json`}
                      render={() => <Json jobRun={jobRun} />}
                    />
                    <Route render={() => <Overview jobRun={jobRun} />} />
                  </Switch>
                </Grid>
              </Grid>
            </Grid>
            <Grid item xs={4}>
              <StatusCard {...jobRun} title={jobRun.status} />
            </Grid>
          </Grid>
        )}
      </Content>
    </>
  )
}

export default Show
