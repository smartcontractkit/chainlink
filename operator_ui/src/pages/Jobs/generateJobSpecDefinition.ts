import { ApiResponse } from '@chainlink/json-api-client'
import { JobSpec, OcrJobSpec } from 'core/store/models'
import { stringifyJobSpec, JobSpecFormats } from './utils'

type DIRECT_REQUEST_DEFINITION_VALID_KEYS =
  | 'name'
  | 'initiators'
  | 'tasks'
  | 'startAt'
  | 'endAt'

const asUnknownObject = (object: object) => object as { [key: string]: unknown }

const scrub = ({
  payload,
  keysToRemove,
}: {
  payload: unknown
  keysToRemove: string[]
}): JSONValue => {
  if (typeof payload === 'string' || payload === null) {
    return payload
  }

  if (Array.isArray(payload)) {
    return payload.map((p) => scrub({ payload: p, keysToRemove }))
  }

  if (typeof payload === 'object' && payload !== null) {
    const typedPayload = asUnknownObject(payload)
    const keepers = Object.keys(typedPayload).filter(
      (k) => !keysToRemove.includes(k),
    )
    return keepers.reduce((accumulator, key) => {
      const value = typedPayload[key]
      if (
        value === null ||
        (typeof value === 'object' &&
          value !== null &&
          Object.keys(value).length === 0)
      ) {
        return accumulator
      }
      return { ...accumulator, [key]: value }
    }, {})
  }

  return null
}

type ScrubbedJobSpec = { [key in DIRECT_REQUEST_DEFINITION_VALID_KEYS]: any }

export const generateJSONDefinition = (
  job: ApiResponse<JobSpec>['data']['attributes'],
): string => {
  const scrubbedJobSpec: ScrubbedJobSpec = ([
    'name',
    'initiators',
    'tasks',
    'startAt',
    'endAt',
  ] as DIRECT_REQUEST_DEFINITION_VALID_KEYS[]).reduce((accumulator, key) => {
    const value = scrub({
      payload: job[key],
      keysToRemove: ['ID', 'CreatedAt', 'DeletedAt', 'UpdatedAt'],
    })

    if (value === null) {
      return accumulator
    }
    return {
      ...accumulator,
      [key]: value,
    }
  }, {} as ScrubbedJobSpec)

  /**
   * We want to remove the name field if it was auto-generated
   * to avoid running into FK constraint errors when duplicating
   * a job spec.
   */
  if (scrubbedJobSpec.name.includes(job.id)) {
    delete scrubbedJobSpec.name
  }

  return stringifyJobSpec({
    value: scrubbedJobSpec,
    format: JobSpecFormats.JSON,
  })
}

export const generateTOMLDefinition = (
  jobSpecAttributes: ApiResponse<OcrJobSpec>['data']['attributes'],
): string => {
  const ocrSpecWithoutDates = {
    ...jobSpecAttributes.offChainReportingOracleSpec,
    createdAt: undefined,
    updatedAt: undefined,
  }

  return stringifyJobSpec({
    value: {
      type: 'offchainreporting',
      schemaVersion: 1,
      ...ocrSpecWithoutDates,
      observationSource: jobSpecAttributes.pipelineSpec.dotDagSource,
    },
    format: JobSpecFormats.TOML,
  })
}
