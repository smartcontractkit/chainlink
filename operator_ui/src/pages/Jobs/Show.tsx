import React from 'react'
import { v2 } from 'api'
import { Route, RouteComponentProps, Switch } from 'react-router-dom'
import { useErrorHandler } from 'hooks/useErrorHandler'
import { useLoadingPlaceholder } from 'hooks/useLoadingPlaceholder'
import {
  generateJSONDefinition,
  generateTOMLDefinition,
} from './generateJobSpecDefinition'
import { PaginatedApiResponse } from '@chainlink/json-api-client'
import { OcrJobRun, RunStatus } from 'core/store/models'
import { JobData } from './sharedTypes'
import { JobsDefinition } from './Definition'
import { JobsErrors } from './Errors'
import { RecentRuns } from './RecentRuns'
import { RegionalNav } from './RegionalNav'
import { Runs as JobRuns } from './Runs'

type Props = RouteComponentProps<{
  jobSpecId: string
}>

const DEFAULT_PAGE = 1
const RECENT_RUNS_COUNT = 5

function getOcrJobStatus({
  attributes: { finishedAt, errors },
}: NonNullable<PaginatedApiResponse<OcrJobRun[]>>['data'][0]) {
  if (finishedAt === null) {
    return RunStatus.IN_PROGRESS
  }
  if (errors[0] !== null) {
    return RunStatus.ERRORED
  }
  return RunStatus.COMPLETED
}

export const JobsShow: React.FC<Props> = ({ match }) => {
  const [state, setState] = React.useState<JobData>({
    recentRuns: [],
    recentRunsCount: 0,
  })
  const { job, jobSpec } = state
  const { error, ErrorComponent, setError } = useErrorHandler()
  const { LoadingPlaceholder } = useLoadingPlaceholder(!error && !jobSpec)

  const { jobSpecId } = match.params
  // `isNaN` actually accepts strings and we don't want to `parseInt` or `parseFloat`
  //  as it doesn't have the behaviour we want.
  const isOcrJob = !isNaN((jobSpecId as unknown) as number)

  const getJobSpecRuns = React.useCallback(
    ({ page = DEFAULT_PAGE, size = RECENT_RUNS_COUNT } = {}) => {
      const requestParams = {
        jobSpecId,
        page,
        size,
      }
      if (isOcrJob) {
        return v2.ocrRuns
          .getJobSpecRuns(requestParams)
          .then((jobSpecRunsResponse) => {
            setState((s) => ({
              ...s,
              recentRuns: jobSpecRunsResponse.data.map((jobRun) => ({
                createdAt: jobRun.attributes.createdAt,
                id: jobRun.id,
                status: getOcrJobStatus(jobRun),
                jobId: jobSpecId,
              })),
              recentRunsCount: jobSpecRunsResponse.meta.count,
            }))
          })
          .catch(setError)
      } else {
        return v2.runs
          .getJobSpecRuns(requestParams)
          .then((jobSpecRunsResponse) => {
            setState((s) => ({
              ...s,
              recentRuns: jobSpecRunsResponse.data.map((jobRun) => ({
                createdAt: jobRun.attributes.createdAt,
                id: jobRun.id,
                status: jobRun.attributes.status,
                jobId: jobSpecId,
              })),
              recentRunsCount: jobSpecRunsResponse.meta.count,
            }))
          })
          .catch(setError)
      }
    },
    [isOcrJob, jobSpecId, setError],
  )

  const getJobSpec = React.useCallback(async () => {
    if (isOcrJob) {
      return v2.ocrSpecs
        .getJobSpec(jobSpecId)
        .then((response) => {
          const jobSpec = response.data
          setState((s) => ({
            ...s,
            jobSpec,
            job: {
              ...jobSpec.attributes.offChainReportingOracleSpec,
              id: jobSpec.id,
              errors: jobSpec.attributes.errors,
              definition: generateTOMLDefinition(jobSpec.attributes),
              type: 'Off-chain reporting',
            },
          }))
        })
        .catch(setError)
    } else {
      return v2.specs
        .getJobSpec(jobSpecId)
        .then((response) => {
          const jobSpec = response.data
          setState((s) => ({
            ...s,
            jobSpec,
            job: {
              ...jobSpec.attributes,
              id: jobSpec.id,
              type: 'Direct request',
              definition: generateJSONDefinition(jobSpec.attributes),
            },
          }))
        })
        .catch(setError)
    }
  }, [isOcrJob, jobSpecId, setError])

  React.useEffect(() => {
    getJobSpec()
  }, [getJobSpec])

  return (
    <div>
      <RegionalNav
        jobSpecId={jobSpecId}
        job={job}
        getJobSpecRuns={getJobSpecRuns}
        runsCount={state.recentRunsCount}
      />
      <Switch>
        <Route
          path={`${match.path}/definition`}
          render={() => (
            <JobsDefinition
              {...{
                ...state,
                ErrorComponent,
                LoadingPlaceholder,
                error,
              }}
            />
          )}
        />
        <Route
          path={`${match.path}/errors`}
          render={() => (
            <JobsErrors
              {...{
                ...state,
                ErrorComponent,
                LoadingPlaceholder,
                error,
                getJobSpec,
                setState,
              }}
            />
          )}
        />
        <Route
          path={`${match.path}/runs`}
          render={() => (
            <JobRuns
              {...{
                ...state,
                error,
                ErrorComponent,
                LoadingPlaceholder,
                getJobSpecRuns,
              }}
            />
          )}
        />
        <Route
          path={`${match.path}`}
          render={() => (
            <RecentRuns
              {...{
                ...state,
                error,
                ErrorComponent,
                LoadingPlaceholder,
                getJobSpecRuns,
              }}
            />
          )}
        />
      </Switch>
    </div>
  )
}

export default JobsShow
