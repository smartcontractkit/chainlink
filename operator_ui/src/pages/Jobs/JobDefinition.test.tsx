import React from 'react'
import { JobsShow } from 'pages/Jobs/Show'
import { Route } from 'react-router-dom'
import { mountWithProviders } from 'test-helpers/mountWithTheme'
import { syncFetch } from 'test-helpers/syncFetch'
import globPath from 'test-helpers/globPath'

import {
  jsonApiJob,
  fluxMonitorJobResource,
} from 'support/factories/jsonApiJobs'

const JOB_ID = '200'

describe('pages/Jobs/Definition', () => {
  it('renders the job definition component', async () => {
    // Mock the job fetch
    const jobResponse = jsonApiJob(
      fluxMonitorJobResource({
        id: JOB_ID,
      }),
    )
    global.fetch.getOnce(globPath(`/v2/jobs/${JOB_ID}`), jobResponse)

    const wrapper = mountWithProviders(
      <Route path="/jobs/:jobId" component={JobsShow} />,
      {
        initialEntries: [`/jobs/${JOB_ID}/definition`],
      },
    )

    await syncFetch(wrapper)

    expect(wrapper.find('code').text()).toMatchSnapshot()
  })
})
