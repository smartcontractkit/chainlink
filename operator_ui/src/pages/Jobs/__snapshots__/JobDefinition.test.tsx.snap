// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Jobs/Definition renders the job definition component 1`] = `
"type = \\"fluxmonitor\\"
schemaVersion = 1
name = \\"Flux Monitor V2 job\\"
maxTaskDuration = \\"\\"
observationSource = \\"\\"\\"
   fetch    [type=http method=POST url=\\"http://localhost:8001\\" requestData=\\"{\\\\\\\\\\"hi\\\\\\\\\\": \\\\\\\\\\"hello\\\\\\\\\\"}\\"];
    parse    [type=jsonparse path=\\"data,result\\"];
    multiply [type=multiply times=100];
    fetch -> parse -> multiply;
\\"\\"\\"
externalJobID = \\"0EEC7E1D-D0D2-476C-A1A8-72DFB6633F47\\"

"
`;
